lib/spack/spack/cmd/testinstall.py

```diff
@@ -0,0 +1,181 @@
+                failure.set('type', "Build Error")
+                failure.text = buildInfo
+        ET.ElementTree(self.root).write(stream)
+
+
+class BuildId(object):
+    def __init__(self, name, version, hashId):
+        self.name = name
+        self.version = version
+        self.hashId = hashId
+    
+    def stringId(self):
+        return "-".join(str(x) for x in (self.name, self.version, self.hashId))
+
+
+def createTestOutput(spec, handled, output):
+    if spec in handled:
+        return handled[spec]
+    
+    childSuccesses = list(createTestOutput(dep, handled, output) 
+            for dep in spec.dependencies.itervalues())
+    package = spack.db.get(spec)
+    handled[spec] = package.installed
+    
+    if all(childSuccesses):
+        bId = BuildId(spec.name, spec.version, spec.dag_hash())
+
+        if package.installed:
+            buildLogPath = spack.install_layout.build_log_path(spec)
+        else:
+            #TODO: search recursively under stage.path instead of only within
+            #    stage.source_path
+            buildLogPath = join_path(package.stage.source_path, 'spack-build.out')            
+
+        with open(buildLogPath, 'rb') as F:
+            buildLog = F.read() #TODO: this may not return all output
+            #TODO: add the whole build log? it could be several thousand
+            #    lines. It may be better to look for errors.
+            output.addTest(bId, package.installed, buildLogPath + '\n' +
+                spec.to_yaml() + buildLog)
+    #TODO: create a failed test if a dependency didn't install?
+
+    return handled[spec]
+
+
+def testinstall(parser, args):
+    if not args.packages:
+        tty.die("install requires at least one package argument")
+
+    if args.jobs is not None:
+        if args.jobs <= 0:
+            tty.die("The -j option must be a positive integer!")
+
+    if args.no_checksum:
+        spack.do_checksum = False        # TODO: remove this global.
+
+    specs = spack.cmd.parse_specs(args.packages, concretize=True)
+    newInstalls = set()
+    for spec in itertools.chain.from_iterable(spec.traverse() 
+            for spec in specs):
+        package = spack.db.get(spec)
+        if not package.installed:
+            newInstalls.add(spec)
+    
+    try:
+        for spec in specs:
+            package = spack.db.get(spec)
+            if not package.installed:
+                package.do_install(
+                    keep_prefix=False,
+                    keep_stage=False,
+                    ignore_deps=False,
+                    make_jobs=args.jobs,
+                    verbose=args.verbose,
+                    fake=False)
+    finally:
+        #TODO: note if multiple packages are specified and a prior one fails, 
+        #    it will prevent any of the others from succeeding even if they
+        #    don't share any dependencies. i.e. the results may be strange if
+        #    you run testinstall with >1 top-level package
```

Comment:
In this case the test for dependents of the failed dependency should be "not completed" ([see here](https://confluence.atlassian.com/bamboo/viewing-a-build-result-289276921.html)).

If you do a post-order traversal, you can check whether prior failed tests are in the deps of later ones, and you can just mark the ones that couldn't be built as not completed.